trantor/net/inner/tlsprovider/BotanTLSProvider.cc

```diff
@@ -0,0 +1,390 @@
+        }
+        else
+        {
+            validationPolicy_.requireClientCert_ =
+                contextPtr_->requireClientCert;
+            channel_ = std::make_unique<Botan::TLS::Client>(
+                *this,
+                sessionManager,
+                *credsPtr_,
+                validationPolicy_,
+                rng,
+                Botan::TLS::Server_Information(policyPtr_->getHostname(),
+                                               conn_->peerAddr().toPort()),
+                policyPtr_->getUseOldTLS()
+                    ? Botan::TLS::Protocol_Version::TLS_V10
+                    : Botan::TLS::Protocol_Version::TLS_V12,
+                policyPtr_->getAlpnProtocols());
+            setSniName(policyPtr_->getHostname());
+        }
+    }
+
+    void handleSSLError(SSLError err)
+    {
+        if (!errorCallback_)
+            return;
+        loop_->queueInLoop([this, err]() { errorCallback_(conn_, err); });
+    }
+
+    virtual ~BotanTLSProvider() override = default;
+
+    void tls_emit_data(const uint8_t data[], size_t size) override
+    {
+        auto n = writeCallback_(conn_, data, size);
+        lastWriteSize_ = n;
+
+        // store the unsent data and send it later
+        if (n == ssize_t(size))
+            return;
+        if (n == -1)
+            n = 0;
+        appendToWriteBuffer((const char *)data + n, size - n);
+    }
+
+    void tls_record_received(uint64_t seq_no,
+                             const uint8_t data[],
+                             size_t size) override
+    {
+        (void)seq_no;
+        recvBuffer_.append((const char *)data, size);
+        if (messageCallback_)
+            messageCallback_(conn_, &recvBuffer_);
+    }
+
+    void tls_alert(Botan::TLS::Alert alert) override
+    {
+        if (alert.type() == Botan::TLS::Alert::CLOSE_NOTIFY)
+        {
+            LOG_WARN << "TLS close notify received";
+            if (closeCallback_)
+                closeCallback_(conn_);
+        }
+        else
+        {
+            if (errorCallback_)
+                errorCallback_(conn_, SSLError::kSSLProtocolError);
+        }
+    }
+
+    bool tls_session_established(const Botan::TLS::Session &session)
+    {
+        LOG_TRACE << "tls_session_established";
+        bool needCerts =
+            policyPtr_->getValidateDate() || policyPtr_->getValidateDomain();
+        const auto &certs = session.peer_certs();
+        if (needCerts && session.peer_certs().empty())
+        {
+            handleSSLError(SSLError::kSSLInvalidCertificate);
+            throw Botan::TLS::TLS_Exception(Botan::TLS::Alert::NO_CERTIFICATE,
+                                            "No certificate received");
+        }
+        if (!certs.empty())
+        {
+            auto &cert = certs[0];
+            if (policyPtr_->getValidateDomain() &&
+                cert.matches_dns_name(policyPtr_->getHostname()) == false)
```

Comment:
I've update the handling to use the normal `tls_verify_cert_chain` when possible. However, I can't find a way to allow self signed certificates while still checking the hostname and certificate dates. This is useful for protocols that rely on TOFU.

For example, Gemini uses TOFU but still validates the hostname to detect server misconfiguration and dates to somewhat force renewal.

>  The first time a Gemini client connects to a server, it accepts whatever certificate it is presented. That certificate's fingerprint and expiry date are saved in a persistent database (like the .known_hosts file for SSH), associated with the server's hostname. ....

Reply:
For the self-signed certificate case, you might try using `Botan::PKIX::check_chain()`. That's a low-level API but might serve you well as it has a special case for self-signed end entities. It will check all the things you implemented in `Callbacks::tls_verify_cert_chain()` and more. I personally didn't test it for your use case, though. Please make sure to test it before taking my word for it. ;) Note as well that it won't check any revocation information for the self-signed cert.

Something along those lines should get you started:

```C++
std::shared_ptr<const X509_Certificate> self_signed_ee_cert = /* the certificate */;

std::vector<std::shared_ptr<const X509_Certificate>> self_signed;
self_signed.push_back(self_signed_ee_cert);

Botan::Path_Validation_Restrictions
    restrictions(false /* require revocation */,
                 tls_policy.minimum_signature_strength());

const auto status = 
    Botan::PKIX::check_chain(self_signed, ref_time,
                             hostname, usage,
                             restrictions.minimum_key_strength(),
                             restrictions.trusted_hashes());

const auto result = Botan::PKIX::overall_status(status);

if(result == Botan::Certificate_Status_Code::OK) {
    // success
}
```

Also, I found the policy checks in `Callbacks::tls_verify_cert_chain()` somewhat surprising. Will comment inline shortly.